### ISSUE TITLE
Package dependencies not installed when private PyPi repository present

### ISSUE THREAD

**Issue by michhri** [python-poetry/poetry#4000]
<!--
  Hi there! Thank you for discovering and submitting an issue.

  Before you submit this; let's make sure of a few things.
  Please make sure the following boxes are ticked if they are correct.
  If not, please try and fulfill these first.
-->

<!-- Checked checkbox should look like this: [x] -->
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

<!--
  Once those are done, if you're able to fill in the following list with your information,
  it'd be very helpful to whoever handles the issue.
-->

- **OS version and name**: MacOS Big Sur 11.2.3
- **Poetry version**: 1.1.6

## Issue
<!-- Now feel free to write your issue, but please be descriptive! Thanks again 🙌 ❤️ -->
I encountered an issue with dependencies when using a private repository. I have a `pyproject.toml` like
```toml
[tool.poetry]
name = "my-awesome-package"
version = "0.1.0"
description = "My package."
authors = ["Me"]
packages = [
    { include = "sub_package_1" },
    { include = "sub_package_2" }
]
include = ["path_to_files/**/*.gz"]

[[tool.poetry.source]]
name = 'private_pypi'
url = 'https://mypypi.dom/pypi/simple'

[tool.poetry.dependencies]
python = "^3.7"

[tool.poetry.dev-dependencies]

[build-system]
build-backend = "poetry.masonry.api"
requires = ["poetry>=1.1.2"]
``` 
And I'd like to add a new package, explicitly `poetry add rasterio`. 
When the private repository (which has nothing to do with the desired package) is present in `pyproject.toml`, only 
the package itself gets installed without its dependencies
```
poetry add rasterio                                                                                                                                                                                                                                                
Using version ^1.2.3 for rasterio

Updating dependencies
Resolving dependencies... (10.2s)

Writing lock file

Package operations: 1 install, 0 updates, 0 removals

  • Installing rasterio (1.2.3)
```
When I remove the private repository part
```toml
[[tool.poetry.source]]
name = 'private_pypi'
url = 'https://mypypi.dom/pypi/simple'
```
and try to add the package, all of its dependencies get installed correctly
```
poetry add rasterio
Using version ^1.2.3 for rasterio

Updating dependencies
Resolving dependencies... (3.5s)

Writing lock file

Package operations: 6 installs, 0 updates, 0 removals

  • Installing click (7.1.2)
  • Installing affine (2.3.0)
  • Installing click-plugins (1.1.1)
  • Installing cligj (0.7.1)
  • Installing snuggs (1.4.7)
  • Installing rasterio (1.2.3)
```
It seems that the private repository is somehow interfering with the resolution of the dependencies. 
I have tried to recreate both the `poetry.lock` file and the virtual environment from scratch with no change in this behaviour.  

Initially, I have discovered this issue through a transitive dependency (a dependency on the private pypi repository, which I was trying to install, depends on `rasterio`) when I tried to `poetry update`, the dependencies `affine`, `click-plugins `, `snuggs` and `cligj` got uninstalled, so the issue seemed to be related to #3224 and #3684, but the issue can be reproduced with this minimalistic setup.

I have specified the missing dependencies explicitly in the `pyproject.toml` file for the time being. This however is a suboptimal solution as it bloats `pyproject.toml`.

Thank you for your help!

**Comment by theirix** [comment]
Reproduced with Poetry 1.2.2 / macOS 12.6.1 (21G217) with exactly this package `rasterio`.

The `pyproject.toml`:
```toml
[tool.poetry]
name = "foobar"
version = "0.1.0"
description = ""
authors = ["theirix <theirix@gmail.com>"]
readme = "README.md"

[tool.poetry.dependencies]
python = "^3.9"
rasterio = "1.2.10"


[build-system]
requires = ["poetry-core"]
build-backend = "poetry.core.masonry.api"

[[tool.poetry.source]]
name = "REDACTED"
url = "REDACTED"
```

Commands:
```
% export REPOSITORY_USERNAME=REDACTED
% export REPOSITORY_PASSWORD=REDACTED
% poetry new foobar
% cd foobar
% # add a private PyPi repository (GitLab-based if it matters) to the toml
% poetry env use /usr/local/opt/python@3.9/bin/python3.9
Using virtualenv: /Users/irix/.virtualenvs/foobar-w9rLp08K-py3.9
% poetry add rasterio@1.2.10

Updating dependencies
Resolving dependencies... (0.1s)

Writing lock file

Package operations: 1 install, 0 updates, 0 removals

  • Installing rasterio (1.2.10)
```

If I don't use a private repository, all dependencies are installed as expected:
```
Package operations: 10 installs, 1 update, 0 removals

  • Installing click (8.1.3)
  • Installing numpy (1.23.5)
  • Installing pyparsing (3.0.9)
  • Installing affine (2.3.1)
  • Installing attrs (22.1.0)
  • Installing certifi (2022.9.24)
  • Installing click-plugins (1.1.1)
  • Installing cligj (0.7.2)
  • Updating setuptools (65.5.1 -> 65.6.3)
  • Installing snuggs (1.4.7)
  • Installing rasterio (1.2.10)
```

**Comment by theirix** [comment]
@neersighted newest 1.3.1 still has this issue. I think this bug should be reopened.

**Comment by neersighted** [comment]
I'm still unable to reproduce this against devpi. Can you please provide `poetry add` in verbose mode, as well as the precise command you're using to add a custom source? How are you communicating the credentials to Poetry? They should be added using `poetry config`.

**Comment by theirix** [comment]
I decided to reread the documentation about [secondary package sources](https://python-poetry.org/docs/repositories/#secondary-package-sources). My scenario was about publishing private packages to the private repository. It is not the same as using the private repository as a full PyPI mirror.

So in my example, I hadn't marked the repository as secondary (no `secondary = true` in my snippet above). So poetry expects to find all packages in a private repository, even those that weren't published to the secondary repository. After marking it as a secondary, poetry successfully installs a `rasterio` package with all its dependencies.

**Comment by mbbyn** [comment]
Hello,

We are facing the same issue, and I'm not sure anyone posted the `-vvv` logs before, so I will add it from our side.

I'm using poetry to install tensorflow 2.11.0 from devpi, however, it is resolving the dependencies differently than if the sources was PyPI directly. I can also provide a full reproduce case with docker-compose, as we use docker for both devpi and poetry.

Here is the basic output below when trying to install tensorflow from PyPI vs. devpi. It seems poetry is installing the tensorflow (intel / aws), causing it to detect circular dependencies, and eventually failing silently.

Using the following versions:
Python 3.10.9
Poetry 1.3.2 or 1.4.0
devpi-server 6.8.0

```
Source (PyPI): 82 packages found for tensorflow *
Using version ^2.11.0 for tensorflow

Updating dependencies
Resolving dependencies...
   1: fact: test is 0.1.0
   1: derived: test
   1: fact: test depends on tensorflow (^2.11.0)
   1: selecting test (0.1.0)
   1: derived: tensorflow (>=2.11.0,<3.0.0)
Source (PyPI): 1 packages found for tensorflow >=2.11.0,<3.0.0
   1: fact: tensorflow (2.11.0) depends on absl-py (>=1.0.0)
   1: fact: tensorflow (2.11.0) depends on astunparse (>=1.6.0)
   1: fact: tensorflow (2.11.0) depends on flatbuffers (>=2.0)
   1: fact: tensorflow (2.11.0) depends on gast (>=0.2.1,<=0.4.0)
   1: fact: tensorflow (2.11.0) depends on google-pasta (>=0.1.1)
   1: fact: tensorflow (2.11.0) depends on h5py (>=2.9.0)
   1: fact: tensorflow (2.11.0) depends on libclang (>=13.0.0)
   1: fact: tensorflow (2.11.0) depends on numpy (>=1.20)
   1: fact: tensorflow (2.11.0) depends on opt-einsum (>=2.3.2)
   1: fact: tensorflow (2.11.0) depends on packaging (*)
   1: fact: tensorflow (2.11.0) depends on protobuf (>=3.9.2,<3.20)
   1: fact: tensorflow (2.11.0) depends on setuptools (*)
   1: fact: tensorflow (2.11.0) depends on six (>=1.12.0)
   1: fact: tensorflow (2.11.0) depends on termcolor (>=1.1.0)
   1: fact: tensorflow (2.11.0) depends on typing-extensions (>=3.6.6)
   1: fact: tensorflow (2.11.0) depends on wrapt (>=1.11.0)
   1: fact: tensorflow (2.11.0) depends on grpcio (>=1.24.3,<2.0)
   1: fact: tensorflow (2.11.0) depends on tensorboard (>=2.11,<2.12)
   1: fact: tensorflow (2.11.0) depends on tensorflow-estimator (>=2.11.0,<2.12)
   1: fact: tensorflow (2.11.0) depends on keras (>=2.11.0,<2.12)
   1: fact: tensorflow (2.11.0) depends on tensorflow-io-gcs-filesystem (>=0.23.1)
   1: selecting tensorflow (2.11.0)
   1: derived: tensorflow-io-gcs-filesystem (>=0.23.1)
   1: derived: keras (>=2.11.0,<2.12)
   1: derived: tensorflow-estimator (>=2.11.0,<2.12)
   1: derived: tensorboard (>=2.11,<2.12)
   1: derived: grpcio (>=1.24.3,<2.0)
   1: derived: wrapt (>=1.11.0)
   1: derived: typing-extensions (>=3.6.6)
   1: derived: termcolor (>=1.1.0)
   1: derived: six (>=1.12.0)
   1: derived: setuptools
   1: derived: protobuf (>=3.9.2,<3.20)
   1: derived: packaging
   1: derived: opt-einsum (>=2.3.2)
   1: derived: numpy (>=1.20)
   1: derived: libclang (>=13.0.0)
   1: derived: h5py (>=2.9.0)
   1: derived: google-pasta (>=0.1.1)
   1: derived: gast (>=0.2.1,<=0.4.0)
   1: derived: flatbuffers (>=2.0)
   1: derived: astunparse (>=1.6.0)
   1: derived: absl-py (>=1.0.0)
```

```
Source (devpi): 82 packages found for tensorflow *
Using version ^2.11.0 for tensorflow

Updating dependencies
Resolving dependencies...
   1: fact: test is 0.1.0
   1: derived: test
   1: fact: test depends on tensorflow (^2.11.0)
   1: selecting test (0.1.0)
   1: derived: tensorflow (>=2.11.0,<3.0.0)
Source (devpi): 1 packages found for tensorflow >=2.11.0,<3.0.0
   1: fact: tensorflow (2.11.0) depends on tensorflow-cpu-aws (2.11.0)
   1: fact: tensorflow (2.11.0) depends on tensorflow-intel (2.11.0)
   1: selecting tensorflow (2.11.0)
   1: derived: tensorflow-intel (==2.11.0)
   1: derived: tensorflow-cpu-aws (==2.11.0)
```

**Comment by mbbyn** [comment]
I tried to inspect it further today, and noticed that the package definition is somehow different based on the source (PyPI vs devpi). Here is a cleaned up diff to see the difference. Notice how the requires for tensorflow is empty from PyPI but has two false positives for devpi.

https://editor.mergely.com/zDeiGtTI/

**Comment by mbbyn** [comment]
### Hacky Solution

Here is a hacky solution for the impatient. Only tested on our linux docker image pipeline.

**Use at your own risk!**

The patch script:

```python
"""
Script to patch poetry to fix fetching certain wheels from legacy repository.
"""

# filter wheels func to be appended at the end of the file
FILTER_WHEELS_FUNC = '''
def filter_wheels(wheels):
    """Filter out wheels that are not compatible with the current platform."""
    import sys
    from poetry.utils.env import SystemEnv

    env = SystemEnv(Path(sys.executable))
    markers = env.get_marker_env()

    filtered_wheels = wheels

    platform_wheels = [w for w in filtered_wheels if markers.get("sys_platform", "") in w]
    if platform_wheels:
        filtered_wheels = platform_wheels

    machine_wheels = [w for w in filtered_wheels if markers.get("platform_machine", "") in w]
    if machine_wheels:
        filtered_wheels = machine_wheels

    
    return filtered_wheels
'''

# The function call to filter the wheels
FILTER_WHEELS_CALL = "first_wheel = filter_wheels(platform_specific_wheels)[0]"

# The code to be replaced by the filter_wheels_call
FILTER_WHEELS_CALL_TO_REPLACE="first_wheel = platform_specific_wheels[0]"


def main():
    import sys
    from pathlib import Path

    # get the path to the poetry package
    poetry_path = Path(sys.argv[1]).resolve()

    # search for the target file to be patched
    filename = "http_repository.py"
    target_file = next(poetry_path.glob(f"**/{filename}"))

    # patch it
    with target_file.open("r") as f:
        content = f.read()
    
    # append the filter_wheels function
    content += "\n\n"
    content += FILTER_WHEELS_FUNC

    # replace the function call
    content = content.replace(FILTER_WHEELS_CALL_TO_REPLACE, FILTER_WHEELS_CALL)

    # write the patched file
    with target_file.open("w") as f:
        f.write(content)
    
    print(f"Successfully patched {target_file}")


if __name__ == "__main__":
    main()

```

In the Dockerfile:

```dockerfile
# patch poetry
# see: https://github.com/python-poetry/poetry/issues/4000
COPY --chown=${USER_ID}:${USER_ID} ./resources/patch_poetry.py /tmp/patch_poetry.py
RUN python /tmp/patch_poetry.py "/home/${USER}/.local/share/pypoetry/venv/lib" \
  && rm /tmp/patch_poetry.py
```

**Comment by dimbleby** [comment]
it's not an optimization, though, because wheels downloaded at this point aren't cached.  So downloading the platform-compatible one at this point wouldn't save a later download.

we discussed this at https://github.com/python-poetry/poetry/pull/6547#issuecomment-1250084271 and took the position that it was better to encourage packagers to publish consistent metadata.  

I'm not 100% against "prefer a platform-compatible wheel": but cross-platform builds are important to poetry and poetry users, and unless metadata is consistent then it can't achieve that.  

I'd definitely recommend that one of you who cares about this at least try submitting an MR to tensorflow!

**Comment by mbbyn** [comment]
@dimbleby TF is a downstream dependency for us, and could take 6 months or more to be resolved on our side. Sure, would be good for the long run, but as app developers (sorry for the cynicism) but we care mostly about our stuff working, rather than the ecosystem as a whole.

Regardless though, appreciate your time and thoughts on the subject!

**Comment by dimbleby** [comment]
not the same thing

I think this one should be closed, it seems to have boiled down to "poetry assumes that all distributions of a package provide the same metadata and goes wrong when that is not true" - which is indeed the case, but I think there is very little chance that poetry will change anything on that front.  

(packages that don't follow this pattern should be considered not poetry-compatible, and if their maintainers care about that then they should provide consistent metadata.)

**Comment by nfantone** [comment]
Not sure about it "not being the same thing" (definitely the same symptoms) — but managed to fix it by following [this comment](https://github.com/python-poetry/poetry/issues/7814#issuecomment-2060608520), which solved it for me.

Essentially:

```sh
poetry self lock
poetry self install
rm poetry.lock
poetry install
```

Seems like a bug of sorts to me.

**Comment by dimbleby** [comment]
still not the same thing, but a duplicate of a dozen others, please stop polluting this unrelated issue
